### ISSUE TITLE
Implement workaround for discovery event concurrency issue

### ISSUE THREAD

**Issue by robrap** [edx/edx-arch-experiments#159]
AC:
- [x] Determine who will implement (Discovery or Arch-BOM) (Answer: Discovery)
- [x] ~~Implement delay tactic described below in the consumer signal handler.~~

Note: Implementation was moved to private ticket: https://2u-internal.atlassian.net/browse/PROD-3098

Context:
* There are cases of discovery ending up with 2 UUIDs for the same course run.
* The working theory is that this happens in the case where a new course run(?) is created from Publisher. Discovery starts creating a record with a UUID for this new course. Before committing, in the same atomic transaction, it calls Studio to create a new course. Studio creates the new course run and then sends the new event on the event bus. The discovery event consumer processes this message, and checks the database to see if a UUID already exists. The theory is that it still hasn't committed the data, and thus doesn't yet exist, so it creates another UUID. At a later point in time, the initial UUID is finally committed.
* Dec 22 was the last occurrence of this issue. Additional logging was added on Dec 26 to help prove out this theory.
* The message processing is running legacy Refresh Course Metadata job code. This code has turned out to have a number of issues:
  * This code makes a request to the ecommerce service in a number of date change cases, and presumably would include be run during these requests.
  * This code makes a salesforce call (in some or all cases?), and this call sometimes takes a very long time.
  * These requests (including the request to Studio) seem to be happening within the atomic transaction, which would cause a rollback in case of issues, but also could leave the database transaction open for a long time.

Next Step:
* Prove out theory of concurrency issue. (in progress)

Possible workarounds:
* Do nothing at all. There is a manual work around, but not sure how simple it is to detect the issue, nor to deal with the workaround.
* Delay processing of the messages. This is a short to medium term hack, but could reduce this issue to near zero. More details below.
* Generally shrink the processing time before the transaction is committed. This _may_ help reduce the chance of an issue, but wouldn't eliminate the chance of a concurrency issue.
* Move request to Studio to a post-commit hook, so the database must be updated before informing Studio. This is on the right track for a longer term solution, but now what if this fails? To do this, should the Studio call instead be replaced by an event in discovery that Studio could consume? This will probably take more effort than the owning team can afford to give right now.
* Add additional data to event that details how the course creation was initiated. Discovery could ignore events that it initiated.

More on the delay processing of messages idea:
* If the signal consumer had access to the timestamp that the event was produced (more later), it could delay processing (if necessary) for some configured amount of time.
  * This separate ticket is related to getting the signal consumer this timestamp: https://github.com/openedx/openedx-events/issues/162. This ticket would unblock this solution. 
* The owning team was proposing adding a 1 minute delay. However, if they can detect and recover from an issue, the delay should be shrunk until the error appears at an acceptably low interval (e.g. once a month, or once every N months, etc.). The longer the interval, the longer all updates will take, which may cause other support issues with people wondering again why their data hasn't been updated.
* If many messages come in at the same time, only the first would be artificially delayed. The rest, if they are already old enough according to the timestamp, would be processed immediately.
* Similarly, if messages were replayed, they should be old enough already where no artificial delay will need to be added.
* For the purposes of this change, _any_ timestamp that approximates when the event was produced would work. This could be when the broker received the event from the producer, or when the producer says the event happened. Worst case, the consumer delays a little when it didn't need to in the case that these get out of sync.

Additional Notes:
* It is not clear who will be implementing the workaround in the consumer code. However, if we agree that the delay is the best path, then this work will be blocked on https://github.com/openedx/openedx-events/issues/162, which Arch-BOM will implement.
* [Link to a private Slack thread](https://twou.slack.com/archives/C04ACDVM6A1/p1671714192370149) regarding this issue.
* [Link to private ticket](https://2u-internal.atlassian.net/browse/PROD-3068) regarding this issue.

**Comment by robrap** [comment]
This ticket is being closed. Implementation will be completed by the owners of Discovery.

FYI: @ormsbee: Since this ticket was created because you wanted to discuss with Discovery.
FYI: @DawoudSheraz: In case Dave ever responds here.
